
#!/usr/bin/env python


PACKAGE = "scout_odom"
NODE = "calibrator"


from dynamic_reconfigure.parameter_generator_catkin import *
import math

# Calibration default values
DEFAULT_CONVERGENCE_STEPS     = 10
DEFAULT_TOLERANCE_ABL    	  = 0.00001
DEFAULT_TOLERANCE_GR          = 0.000001
DEFAULT_ONLINE_CALIB          = False

# Robot geometry default values
DEFAULT_WHEEL_RADIUS   		  = 0.1575
DEFAULT_GEAR_RATIO_MAX        = 0.02857142857 
DEFAULT_GEAR_RATIO_MIN        = 0.025

DEFAULT_REAL_BASELINE		  = 0.583
# Insert calibration results if the procedure is done (required to rebuild)
DEFAULT_BASELINE		      = DEFAULT_REAL_BASELINE # without calib -> 0.583
DEFAULT_GEAR_RATIO 		      = (DEFAULT_GEAR_RATIO_MAX + DEFAULT_GEAR_RATIO_MIN) / 2.0 # without calib -> (DEFAULT_GEAR_RATIO_MAX + DEFAULT_GEAR_RATIO_MIN) / 2.0
DEFAULT_CALIB_DONE		      = False # put True if calibration done otherwise False

gen = ParameterGenerator()

# Calibration parameters
gen.add("conv_steps", int_t, 0, "Number of time steps in which convergence condition holds", default=DEFAULT_CONVERGENCE_STEPS)
gen.add("gr_error_tolerance", double_t, 1, "Max MSE considered as convergence in gear ratio calibration", default=DEFAULT_TOLERANCE_GR)
gen.add("abl_error_tolerance", double_t, 2, "Max MSE considered as convergence in apparent baseline calibration", default=DEFAULT_TOLERANCE_ABL)
gen.add("default_abl", double_t, 3, "Apparent baseline starting value", default=DEFAULT_BASELINE)
gen.add("default_gr", double_t, 4, "Gear ratio starting value", default=DEFAULT_GEAR_RATIO)
gen.add("online_calib", bool_t, 5, "Online calibration", default=DEFAULT_ONLINE_CALIB)
gen.add("calib_done", bool_t, 6, "Calibration already performed (default values are calibrated)", default=DEFAULT_CALIB_DONE)
# Robot geometry parameters
gen.add("wheel_radius", double_t, 7, "Radius of robot wheels", default=DEFAULT_WHEEL_RADIUS)
gen.add("gear_ratio_max", double_t, 8, "Max gear ratio value provided by the manufacturer", default=DEFAULT_GEAR_RATIO_MAX)
gen.add("gear_ratio_min", double_t, 9, "Min gear ratio value provided by the manufacturer", default=DEFAULT_GEAR_RATIO_MIN)
gen.add("real_baseline", double_t, 10, "Baseline value provided by the manufacturer", default=DEFAULT_REAL_BASELINE)


exit(gen.generate(PACKAGE, NODE, "calibParams"))